### PR TITLE
Add subchannel test

### PR DESCRIPTION
**Pull request by matthiasgeihs**
Adds sub-channel tests. Requires a fix to go-perun. Draft until the corresponding PR https://github.com/hyperledger-labs/go-perun/pull/359 is merged there.

Relates to #20

### CODE REVIEW

File: channel/conclude.go [hunk 1]
```diff
@@ -98,6 +98,70 @@ func (a *Adjudicator) ensureConcluded(ctx context.Context, req channel.Adjudicat
...
+	req channel.AdjudicatorReq,
+	subStates channel.StateMap,
+) error {
+	states := channel.MakeStateMap()
+	states.Add(req.Tx.State)
+	for _, v := range subStates {
+		states.Add(v)
+	}
+
+	events := make(chan *subscription.Event, adjEventBuffSize)
```
Comment (cryptphil) [inline channel/conclude.go:114]: Maybe add a comment here why we need to subscribe for specific events at this point.

File: client/init_test.go [hunk 2]
```diff
@@ -23,7 +23,7 @@ import (
 )
 
 func init() {
-	plogrus.Set(logrus.WarnLevel, &logrus.TextFormatter{ForceColors: true})
+	plogrus.Set(logrus.DebugLevel, &logrus.TextFormatter{ForceColors: true})
```
Comment (cryptphil) [inline client/init_test.go:26]: By purpose included in this commit?

File: channel/conclude.go (hunk 1 above)
Comment (cryptphil) [inline channel/conclude.go:114]: Maybe add a comment here why we need to subscribe for specific events at this point.

File: client/subchannel_test.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+func TestSubChannelHappy(t *testing.T) {
+	rng := pkgtest.Prng(t)
+	const A, B = 0, 1 // Indices of clients.
+	s := ethchanneltest.NewSetup(t, rng, 2, ethclienttest.BlockInterval, TxFinalityDepth)
+	setups := ethclienttest.MakeRoleSetups(rng, s, [2]string{"Susie", "Tim"})
+	roles := [2]clienttest.Executer{
+		clienttest.NewSusie(t, setups[A]),
+		clienttest.NewTim(t, setups[B]),
+	}
+	// enable stages synchronization
```
Comment (cryptphil) [inline client/subchannel_test.go:27]:
```suggestion
	// Enable stages synchronization.
```

File: client/subchannel_test.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	rng := pkgtest.Prng(t)
+
+	const A, B = 0, 1 // Indices of clients.
+	s := ethchanneltest.NewSetup(t, rng, 2, ethclienttest.BlockInterval, TxFinalityDepth)
+	setups := ethclienttest.MakeRoleSetups(rng, s, [2]string{"DisputeSusie", "DisputeTim"})
+	roles := [2]clienttest.Executer{
+		clienttest.NewDisputeSusie(t, setups[A]),
+		clienttest.NewDisputeTim(t, setups[B]),
+	}
+	// enable stages synchronization
```
Comment (cryptphil) [inline client/subchannel_test.go:70]:
```suggestion
	// Enable stages synchronization.
```

File: client/subchannel_test.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	setups := ethclienttest.MakeRoleSetups(rng, s, [2]string{"Susie", "Tim"})
+	roles := [2]clienttest.Executer{
+		clienttest.NewSusie(t, setups[A]),
+		clienttest.NewTim(t, setups[B]),
+	}
+	// enable stages synchronization
+	stages := roles[A].EnableStages()
+	roles[B].SetStages(stages)
+
+	cfg := clienttest.NewSusieTimExecConfig(
```
Comment (cryptphil) [inline client/subchannel_test.go:31]: This is quite a large config, maybe use consts/variables for single parameters to make it more readable.

**Review by cryptphil — CHANGES_REQUESTED** [review]
Looks good overall, just some small suggestions.

File: client/init_test.go (hunk 2 above)
Comment (matthiasgeihs) [reply client/init_test.go:26]: On purpose, so that we see error in the CI.

File: client/subchannel_test.go (hunk 3 above)
Comment (matthiasgeihs) [reply client/subchannel_test.go:27]: done `// Enable stage synchronization.`

File: client/subchannel_test.go (hunk 5 above)
Comment (matthiasgeihs) [reply client/subchannel_test.go:31]: Done

**Review by cryptphil — APPROVED** [review]

File: channel/funder.go [hunk 6]
```diff
@@ -186,6 +188,21 @@ func (f *Funder) Fund(ctx context.Context, request channel.FundingReq) error {
...
+	timeout, err := NewBlockTimeoutDuration(ctx, f.ContractInterface, req.Params.ChallengeDuration)
+	if err != nil {
+		return nil, nil, errors.WithMessage(err, "creating block timeout")
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	go func() {
+		if err := timeout.Wait(ctx); err != nil && !pcontext.IsContextError(err) {
+			f.log.Warn("Fund: BlockTimeout.Wait runtime error: ", err)
+		}
+		cancel() // cancel funding context on funding timeout
```
Comment (cryptphil) [inline channel/funder.go:201]:
```suggestion
		cancel() // Cancel funding context on funding timeout.
```

File: client/payment_test.go [hunk 7]
```diff
@@ -98,3 +104,54 @@ func TestHappyAliceBob(t *testing.T) {
 
 	log.Info("Happy test done")
 }
+
```
Comment (cryptphil) [inline client/payment_test.go:107]: The comments in this file are missing a period at the end of the sentence.

File: client/subchannel_test.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+	roles[B].SetStages(stages)
+
+	// Build configuration.
+	baseCfg := clienttest.MakeBaseExecConfig(
+		[2]wire.Address{setups[A].Identity.Address(), setups[B].Identity.Address()},
+		s.Asset,
+		[2]*big.Int{big.NewInt(100), big.NewInt(100)},
+		client.WithoutApp(),
+	)
+	const (
```
Comment (cryptphil) [inline client/subchannel_test.go:52]: Looks more readable if the constants and variables are defined in the beginning of the method.

**Review by cryptphil — CHANGES_REQUESTED** [review]

File: channel/funder.go (hunk 6 above)
Comment (matthiasgeihs) [reply channel/funder.go:201]: The code was just moved. The intention is not to cleanup existing code here. Just to satisfy the linter with regards to function length.

File: client/payment_test.go (hunk 7 above)
Comment (matthiasgeihs) [reply client/payment_test.go:107]: The code was just moved. The intention is not to cleanup existing code here. Just to satisfy the linter with regards to function length.

File: client/subchannel_test.go (hunk 8 above)
Comment (matthiasgeihs) [reply client/subchannel_test.go:52]: I think it's a matter of taste. Also, this part did not change since you approved last time.

**Review by cryptphil — APPROVED** [review]

**Pull request merged**